### ISSUE TITLE
possibly use MLX for MacOS users with WhisperSpeech

### ISSUE THREAD

**Issue by BBC-Esq** [collabora/WhisperSpeech#111]
The purpose is the discuss possibly implementing MLX support for MacOS users.  For example, currently Pytorch [doesn't support](https://github.com/pytorch/pytorch/issues/78044) the FFT operation whereas [MLX does](https://ml-explore.github.io/mlx/build/html/python/fft.html).  This means that WhisperSpeech must put certain models and/or tensors etc. on CPU for MacOs users...whereas CUDA users have complete speedup.

Option 1 - implement MLX just where MPS can't be used
Option 2 - completely replace MPS with MLX
Option 3 - replace MPS with MLX as much as possible based on the multiple models involved in WhisperSpeech and whether each specifically can be run with MLX.
Option 4 - offer MLX IN ADDITION to MPS for all MacOs users.

In general, MLX provides a 2-3x speedup compared to MPS across the board [in most cases](https://towardsdatascience.com/how-fast-is-mlx-a-comprehensive-benchmark-on-8-apple-silicon-chips-and-4-cuda-gpus-378a0ae356a0).

Here are some snippets from the Medium article:

<details><summary>Benchmark Setup</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/0d08be9b-75a4-4761-b8ba-9b7dbf5e9415)

</details>

<details><summary>Linear Layer</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/47c49246-beeb-40e3-8f29-cd4828ff24d4)

</details>

<details><summary>Softmax</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/291ee76f-2ba3-4a9f-ab27-a0d24eb884d6)

</details>

<details><summary>Sigmoid</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/d9d15802-1dcc-4792-b014-41d72b1c54aa)

</details>

<details><summary>Concatenation</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/f357425f-1bd0-4571-931b-5a9bc5da27de)

</details>

<details><summary>Binary Cross Entropy</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/7871ad9a-ce86-4981-bdc1-b5a2770808d5)

</details>

<details><summary>Sort</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/86e8bd16-da86-412c-8fb8-500a76503728)

</details>

<details><summary>Conv2D</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/2d3e1f06-a500-4e4b-bcde-ec13ec37ccd8)

</details>

<details><summary>Unified Memory Gamechanger</summary>

![image](https://github.com/collabora/WhisperSpeech/assets/108230321/8712e6ae-0588-4a8e-91e4-c3b48b77bc1d)

</details>

**Comment by signalprime** [comment]
Great initiative @BBC-Esq !  I'll definitely circle back to this one as soon as possible

**Comment by BBC-Esq** [comment]
@signalprime It would take someone with more programming experience than me to implement, especially since I don't own a Mac, but thought I'd start the discussion anyways.  Interested as always in what you find out.

**Comment by BBC-Esq** [comment]
UPDATE:  Looks like Pytorch might be getting support sooner than later...

https://github.com/pytorch/pytorch/commit/53bfae2c066fcd06784dfa051cd7e2eb5ba5c8fa

**Comment by signalprime** [comment]
I'm definitely looking into it.  Reviewing the Vocos model today

**Comment by BBC-Esq** [comment]
> I'm definitely looking into it. Reviewing the Vocos model today

I'd love to learn if you want to keep me posted and teach me along the way, just FYI.  This is not my profession but a hobby.

**Comment by signalprime** [comment]
Absolutely @BBC-Esq, I will keep you in the loop about it.  MLX mimics the pytorch API in most ways.  I've been building models before we had frameworks like TF and Torch, and in this case I'm looking at building the Vocos model using the MLX library.  It just depends on time constraints.  

I recently finished a long project with ML/RL in the finance domain and put in an application with Collabora last week.  Would you put in a nice word for me @jpc?

**Comment by signalprime** [comment]
I'm getting closer.. almost reached the end of the hole.  We have a standard whisper model for MLX already established.  

I was able to convert the Vocos model and weights to MLX, however MLX doesn't have `weight_norm` established yet.  I've dug into the code, and debating when I have time to add the `_weight_norm` primitive to the C++ MLX library

https://github.com/pytorch/pytorch/blob/834c7a1d3ea07878ad87d127ee28606fc140b552/aten/src/ATen/native/WeightNorm.cpp#L50

I'd like to do a little more research before trying that because it could perhaps be handled another way, or not needed at all, kinda like a quick initial pass-through

**Comment by BBC-Esq** [comment]
Interesting...

**Comment by signalprime** [comment]
Good thing I waited. I got a response that it should be possible [using existing ops](https://github.com/ml-explore/mlx-examples/issues/206#issuecomment-1962829362).  

Here is the [whisper model in MLX](https://github.com/ml-explore/mlx-examples/blob/47dd6bd17f3cc7ef95672ea16e443e58ce5eb1bf/whisper/whisper/whisper.py) format, which is used during voice cloning.  

I was working with MLX conversions for all the parts of the [Vocos model](https://github.com/gemelo-ai/vocos/blob/c859e3b7b534f3776a357983029d34170ddd6fc3/vocos/pretrained.py#L9).  Transferring weights wasn't an issue, but components used in the functions also likely need to be updated.  I'm still becoming familiar, but it seems parts can be mixed and matched.. as in a tensor can be converted to an MX array and passed to a MX component and back to a tensor later.  That would appear necessary since I wouldn't want to keep going further and further into torchaudio for example.  Ideally we just put a replacement for the components where torch doesn't yet support the ops.

**Comment by BBC-Esq** [comment]
> Good thing I waited. I got a response that it should be possible [using existing ops](https://github.com/ml-explore/mlx-examples/issues/206#issuecomment-1962829362).
> 
> Here is the [whisper model in MLX](https://github.com/ml-explore/mlx-examples/blob/47dd6bd17f3cc7ef95672ea16e443e58ce5eb1bf/whisper/whisper/whisper.py) format, which is used during voice cloning.
> 
> I was working with MLX conversions for all the parts of the [Vocos model](https://github.com/gemelo-ai/vocos/blob/c859e3b7b534f3776a357983029d34170ddd6fc3/vocos/pretrained.py#L9). Transferring weights wasn't an issue, but components used in the functions also likely need to be updated. I'm still becoming familiar, but it seems parts can be mixed and matched.. as in a tensor can be converted to an MX array and passed to a MX component and back to a tensor later. That would appear necessary since I wouldn't want to keep going further and further into torchaudio for example. Ideally we just put a replacement for the components where torch doesn't yet support the ops.

That's what my intuition was telling me based on what I read about MLX, but I am far from an expert and would have not way to verify it.  My initial hypothesis was that it might be possible to use MLX for some (but not all) of the necessary operations (or whatever you call them), kind of mix and match like you were saying.  Math is math...but again, this it totally a notice intuition kind of thing.

Let me know if I can help out any...

**Comment by BBC-Esq** [comment]
Not sure if it's relevant, but apparently aten::Lupsample_linear1d has been implemented on pytorch's working version (not included in a release yet though):

aten::upsample_linear1d on mps

**Comment by BBC-Esq** [comment]
@signalprime how's it going?  Any updates?

**Comment by signalprime** [comment]
Hi @BBC-Esq I haven't had an opportunity to resume work on this unfortunately, my friend

**Comment by BBC-Esq** [comment]
Hey @signalprime I hope you don't stop working on this kind of stuff even if you don't get the job with Collabora.  I enjoy working with ya and look forward to improving this all-around kick ass library.  Just throwing that out there!

**Comment by jpc** [comment]
@signalprime Sure, I'll see what I can do :)

**Comment by signalprime** [comment]
@jpc yes absolutely, I sent you an email with details.  Looking forward to it!

**Comment by touhi99** [comment]
is it still working with MPS? i couldn't make it run the current main branch, its use CPU only.

**Comment by BBC-Esq** [comment]
@signalprime are you still interested in working on this or whisperspeech in general?
